assets/src/edit-story/app/media/utils/useMediaUploadQueue.js

```diff
@@ -0,0 +1,356 @@
+};
+
+function useMediaUploadQueue() {
+  const {
+    actions: { uploadFile },
+  } = useUploader();
+  const {
+    isFeatureEnabled,
+    isTranscodingEnabled,
+    canTranscodeFile,
+    transcodeVideo,
+  } = useTranscodeVideo();
+
+  const [state, actions] = useReduction(initialState, reducer);
+  const {
+    startUploading,
+    finishUploading,
+    cancelUploading,
+    startTranscoding,
+    finishTranscoding,
+    replacePlaceholderResource,
+  } = actions;
+
+  // TODO: add effect to generate poster via ffmpeg in case resource lacks dimensions.
+  // See https://github.com/google/web-stories-wp/issues/6399
+
+  // Try to update placeholder resources for freshly transcoded file.
+  useEffect(() => {
+    async function updateItems() {
+      await Promise.all(
+        state.queue.map(async (item) => {
+          const { id, file, state: itemState, resource } = item;
+          if (itemState !== 'TRANSCODED' || !resource.isPlaceholder) {
+            return;
+          }
+
+          try {
+            const newResource = await getResourceFromLocalFile(file);
+            replacePlaceholderResource({ id, resource: newResource });
+          } catch {
+            // Not interested in errors here.
+          }
+        })
+      );
+    }
+
+    updateItems();
+  }, [state.queue, replacePlaceholderResource]);
+
+  // Upload files to server, optionally first transcoding them.
+  useEffect(() => {
+    async function uploadItems() {
+      await Promise.all(
+        state.queue.map(async (item) => {
+          const { id, file, state: itemState } = item;
+          if ('PENDING' !== itemState) {
+            return;
+          }
+
+          if (
+            !isFeatureEnabled ||
+            !isTranscodingEnabled ||
+            !canTranscodeFile(file)
+          ) {
+            // TODO: If !isTranscodingEnabled && canTranscodeFile(), tell user to enable transcoding.
+
+            // If transcoding is not enabled, just upload the file normally without any transcoding.
+
+            startUploading({ id });
+
+            trackEvent('upload_media', {
+              file_size: file.size,
+              file_type: file.type,
+            });
+
+            const trackTiming = getTimeTracker('load_upload_media');
+
+            try {
+              const attachment = await uploadFile(file);
+              finishUploading({
+                id,
+                resource: getResourceFromAttachment(attachment),
+              });
+            } catch (error) {
+              // Cancel uploading if there were any errors.
+              cancelUploading({ id, error });
+
+              trackError('upload_media', error?.message);
+            } finally {
+              trackTiming();
+            }
+
+            return;
+          }
+
+          // TODO: Only transcode & optimize video if needed (criteria TBD).
+          // Probably need to use FFmpeg first to get more information (dimensions, fps, etc.)
+          startTranscoding({ id });
```

Comment:
Should we track the how long it takes to transcode a file. Were we not tracking this before?

Reply:
This is done in `useTranscodeVideo` directly now, as mentioned in the PR description.